### PR TITLE
use Expand-Archive and Compress-Archive in windows os utils

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Work towards removal of package:archive and ideally more stable unzipping of artifacts. These commands are available in Powershell 5+, which we already require for windows.

https://github.com/flutter/flutter/issues/45278

Related to https://github.com/flutter/flutter/issues/48014

### CODE REVIEW

**Comment by jonahwilliams** [comment]
FYI @zanderso , this is the opposite direction we initially discussed going in, but it seems like most of the zip crashes are on windows where the tool exclusively uses package:archive.

Since this functionality is already used by the dart download logic, it seems like a good place to start.

File: packages/flutter_tools/lib/src/base/os.dart [hunk 1]
```diff
@@ -315,23 +323,26 @@ class _WindowsUtils extends OperatingSystemUtils {
...
   void unzip(File file, Directory targetDirectory) {
-    final Archive archive = ZipDecoder().decodeBytes(file.readAsBytesSync());
-    _unpackArchive(archive, targetDirectory);
+    _processManager.runSync(<String>[
+      _activePowershell,
+      'Expand-Archive',
+      '-Path',
+      file.path,
+      '-DestinationPath',
+      targetDirectory.path,
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/os.dart:344]: Should we pass `-Force`? I can't tell from the docs https://docs.microsoft.com/de-de/powershell/module/microsoft.powershell.archive/Expand-Archive?view=powershell-7

File: packages/flutter_tools/lib/src/base/os.dart [hunk 2]
```diff
@@ -315,23 +323,26 @@ class _WindowsUtils extends OperatingSystemUtils {
...
+      '-DestinationPath',
+      zipFile.path,
+    ]);
   }
 
   @override
   void unzip(File file, Directory targetDirectory) {
-    final Archive archive = ZipDecoder().decodeBytes(file.readAsBytesSync());
-    _unpackArchive(archive, targetDirectory);
+    _processManager.runSync(<String>[
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/os.dart:338]: The Unix versions throw an exception if the process exits with a non-zero status. Two questions:
1. Should we spawn these the same way with `throwOnError: true`.
2. If the expansion fails, will the powershell process fail with an error?

File: packages/flutter_tools/lib/src/base/os.dart [hunk 3]
```diff
@@ -315,23 +323,26 @@ class _WindowsUtils extends OperatingSystemUtils {
...
+      _activePowershell,
+      'Expand-Archive',
+      '-Path',
+      file.path,
+      '-DestinationPath',
+      targetDirectory.path,
+    ]);
   }
 
   @override
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/os.dart:348]: If `Expand-Archive` fails when the zip file is malformed, I wonder if `verifyZip` could be implemented by using the Windows equivalent of `/dev/null` as the output path.

File: packages/flutter_tools/lib/src/base/os.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/os.dart:344]: From my testing locally it didn't look necessary. It is also not used in https://github.com/flutter/flutter/blob/master/bin/internal/update_dart_sdk.ps1#L81

File: packages/flutter_tools/lib/src/base/os.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/os.dart:338]: Good question, I'll need to test this locally to verify.

File: packages/flutter_tools/lib/src/base/os.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/os.dart:348]: TIL about `NUL` https://stackoverflow.com/questions/313111/is-there-a-dev-null-on-windows

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/os.dart:348]: Looking at cache.dart, I'm not really sure why we're verifying the zip. Expand-Archive doesn't extract anything if it fails, so verification seems like we would just be extracting it twice

File: packages/flutter_tools/lib/src/base/os.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/os.dart:338]: The powershell process exits with 0, but we can check that the stderr output is non-empty

**Comment by jonahwilliams** [comment]
thanks once again for spaces in the CI path lol

**Review by zanderso — COMMENTED** [review]
This looks good. Do you have some sense for whether folks are successfully using cmd.ede or gitbash even though it's not supported officially?

**Comment by jonahwilliams** [comment]
cmd should still work just as well today, but we still require Powershell to download the dart SDK so that is unchanged.

We could make WSL work by verifying we treat it as a linux platform. For git bash we would need to rewrite bin/flutter, but otherwise Process.run should still use cmd, so I can't imagine it making it worse

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**